js/modules/k6/ws/ws.go

```diff
@@ -0,0 +1,362 @@
+	if handler, ok := goja.AssertFunction(handler); ok {
+		s.eventHandlers[event] = append(s.eventHandlers[event], handler)
+	}
+}
+
+func (s *Socket) handleEvent(event string, args ...goja.Value) {
+	if handlers, ok := s.eventHandlers[event]; ok {
+		for _, handler := range handlers {
+			handler(goja.Undefined(), args...)
+		}
+	}
+}
+
+func (s *Socket) Send(message string) {
+	// NOTE: No binary message support for the time being since goja doesn't
+	// support typed arrays.
+	rt := common.GetRuntime(s.ctx)
+
+	writeData := []byte(message)
+	if err := s.conn.WriteMessage(websocket.TextMessage, writeData); err != nil {
+		s.handleEvent("error", rt.ToValue(err))
+	}
+}
+
+func (s *Socket) Ping() {
+	rt := common.GetRuntime(s.ctx)
+	deadline := time.Now().Add(writeWait)
+
+	err := s.conn.WriteControl(websocket.PingMessage, []byte{}, deadline)
+	if err != nil {
+		s.handleEvent("error", rt.ToValue(err))
+	}
+}
+
+func (s *Socket) SetTimeout(fn goja.Callable, timeoutMs int) {
+	// Starts a goroutine, blocks once on the timeout and pushes the callable
+	// back to the main loop through the scheduled channel
+	go func() {
+		select {
+		case <-time.After(time.Duration(timeoutMs) * time.Millisecond):
+			s.scheduled <- fn
+
+		case <-s.done:
+			return
+		}
+	}()
+}
+
+func (s *Socket) SetInterval(fn goja.Callable, intervalMs int) {
+	// Starts a goroutine, blocks forever on the ticker and pushes the callable
+	// back to the main loop through the scheduled channel
+	go func() {
+		ticker := time.NewTicker(time.Duration(intervalMs) * time.Millisecond)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-ticker.C:
+				s.scheduled <- fn
+
+			case <-s.done:
+				return
+			}
+		}
+	}()
+}
+
+func (s *Socket) Close(args ...goja.Value) {
+	code := websocket.CloseGoingAway
+	if len(args) > 0 && !goja.IsUndefined(args[0]) && !goja.IsNull(args[0]) {
+		code = int(args[0].ToInteger())
+	}
+
+	s.closeConnection(code)
+}
+
+func (s *Socket) closeConnection(code int) error {
+	// Attempts to close the websocket gracefully
+
+	var err error
+	s.shutdownOnce.Do(func() {
+		rt := common.GetRuntime(s.ctx)
+
+		err := s.conn.WriteControl(websocket.CloseMessage,
+			websocket.FormatCloseMessage(code, ""),
+			time.Now().Add(writeWait),
+		)
+		if err != nil {
+			s.handleEvent("error", rt.ToValue(err))
+			// Just call the handler, we'll try to close the connection anyway
+		}
+		s.conn.Close()
+
+		// Stops the main control loop
+		close(s.done)
+	})
+
+	return err
+}
+
+// Wraps conn.ReadMessage in a channel
+func readPump(conn *websocket.Conn, readChan chan []byte, errorChan chan error) {
+	defer conn.Close()
+
+	for {
+		_, message, err := conn.ReadMessage()
+		if err != nil {
+			// Only emit the error if we didn't close the socket ourselves
+			if websocket.IsUnexpectedCloseError(err, websocket.CloseGoingAway) {
+				errorChan <- err
+			}
+
+			return
+		}
+
+		message = bytes.TrimSpace(bytes.Replace(message, newline, space, -1))
```

Comment:
Is it really wise to implicitly mangle payloads like this? Do other websocket libraries do it?